model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/entity/JpaClientEntity.java

```diff
@@ -87,6 +89,10 @@ public JpaClientEntity() {
         this.metadata = new JpaClientMetadata();
     }
 
+    public JpaClientEntity(DeepCloner cloner) {
+        this.metadata = new JpaClientMetadata(cloner);
+    }
```

Comment:
Do we plan to instantiate `JpaClientEntity` using this constructor? Right now it's not used from anywhere.

Reply:
It is used from cloner. See: https://github.com/keycloak/keycloak/blob/33ebbdf2c088f24dc78ac80bbc2bfc0e625e213a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/client/JpaClientMapStorageProviderFactory.java#L61.

But I am not sure whether it makes sense to have the cloner stored in the metadata.

Let me summarize how it works for CHM and HR, probably you already know but to be on same page:
1. Each factory has a CLONER created which knows how to create storage-specific entities. (like. `HotRodClientEntity`, `MapClientEntityImpl`, `JpaClientEntity`)
2. When storage obtains an entity it always makes a copy/clone so that it is sure the stored object has the correct type
3. Now, some entities have also nested entities and we need to be able to copy also those, therefore entities (like `MapClientEntityImpl`) store the cloner reference that created the object so when, for example, `setProtocolMapper` is called the entity setter can create a copy of correct type.

Before this change, we had two types of constructors in DeepCloner: parameter-less and constructor with `DeepCloner`. When a class didn't have `DeepCloner` constructor (it has no mutable fields) we used parameter-less for creating it. However, this can cause problems in the future as some entities may be extended with some complex field that requires deep cloning, however, in `DeepCloner` we still use parameter-less constructor, therefore only `DUMB_CLONER` will be used. This issue can be hard to find so we decided to always use constructor with `DeepCloner` inside.

Reply:
Thank you very much for the reply. I wrongly assumed that `JpaClientEntity::new` is equivalent  to `() -> new JpaClientEntity()`. Now when I played a little bit with it I can see what you mean. It is called exactly from [there](https://github.com/keycloak/keycloak/blob/6383f073c26dc2f3ca53a605a74b67ec2cee16f2/model/map/src/main/java/org/keycloak/models/map/common/DeepCloner.java#L392).

`DeepCloner` object is not stored in database. It's ignored when serialization is done. There is the configuration: https://github.com/keycloak/keycloak/blob/6383f073c26dc2f3ca53a605a74b67ec2cee16f2/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/jsonb/JsonbType.java#L73

I'd say that having the cloner in metadata makes sense for cases when we would want to call `deepClone` method on metadata object. I'm not sure if a such situation could occur. But to be on the safe site I think we should leave it as you have it atm.